File: es4/es4.py
#!/usr/bin/env python3

from operator import itemgetter, attrgetter

import re

events = []
pattern = re.compile(r"^\[(\d+)-(\d+)-(\d+) (\d+):(\d+)\] (.*)$")
begin = re.compile(r"^Guard #(\d+) begins shift$")
asleep = re.compile(r"^falls asleep$")
wakeup = re.compile(r"^wakes up$")
with open("input.txt", "r") as f:
    for line in f:
        m = pattern.match(line)
        if m is None:
            continue

        etype = 0
        guard_id = None
        t = begin.match(m.group(6))
        if t:
            guard_id = int(t.group(1))
            etype = 0

        t = asleep.match(m.group(6))
        if t:
            etype = 1

        t = wakeup.match(m.group(6))
        if t:
            etype = 2

        events.append((
            int(m.group(1)),
            int(m.group(2)),
            int(m.group(3)),
            int(m.group(4)),
            int(m.group(5)),
            etype,
            guard_id
        ))

# order events by date
events.sort(key = itemgetter(0,1,2,3,4))

class Guard(object):
    def __init__(self, guard_id):
        self.id = guard_id
        self.minutes = [0 for x in range(60)]
        self.total_asleep = 0
        self.max_asleep = -1
        self.max_asleep_minute = -1

    def compute_asleep(self):
        self.total_asleep = 0
        self.max_asleep = -1
        self.max_asleep_minute = -1
        for i in range(len(self.minutes)):
            self.total_asleep += self.minutes[i]
            if self.max_asleep < self.minutes[i]:
                self.max_asleep = self.minutes[i]
                self.max_asleep_minute = i

    def fill(self, minute, end):
        for i in range(minute, end):
            self.minutes[i] += 1

guard_map = {}

guard = None
state = None
minute = None
for e in events:
    if e[5] == 0:
        # Begin shift
        if state == 1 and guard != e[6]:
            guard.fill(minute, 60)

        guard = guard_map.get(e[6])
        if guard is None:
            guard = Guard(e[6])
            guard_map[e[6]] = guard

    elif e[5] == 1:
        # Falls asleep
        minute = e[4]
    elif e[5] == 2:
        # Wakes up
        guard.fill(minute, e[4])

guards = []
for key, guard in guard_map.items():
    guard.compute_asleep()
    guards.append(guard)

guards.sort(key = attrgetter("total_asleep"), reverse = True)
print("Strategy1: {}".format(guards[0].id * guards[0].max_asleep_minute))

guards.sort(key = attrgetter("max_asleep"), reverse = True)
print("Strategy2: {}".format(guards[0].id * guards[0].max_asleep_minute))

File: es4/es4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type EventType int

const (
	BEGINS_SHIFT EventType = iota
	FALLS_ASLEEP
	WAKES_UP
)

type Event struct {
	Year    int
	Month   int
	Day     int
	Hour    int
	Minute  int
	Type    EventType
	GuardId int
}

func GetEvent(line string) (Event, error) {
	var e Event
	_, err := fmt.Sscanf(line, "[%d-%d-%d %d:%d] Guard #%d begins shift",
		&e.Year, &e.Month, &e.Day, &e.Hour, &e.Minute, &e.GuardId)
	if err == nil {
		e.Type = BEGINS_SHIFT
		return e, nil
	}

	_, err = fmt.Sscanf(line, "[%d-%d-%d %d:%d] falls asleep",
		&e.Year, &e.Month, &e.Day, &e.Hour, &e.Minute)
	if err == nil {
		e.GuardId = 0
		e.Type = FALLS_ASLEEP
		return e, nil
	}

	_, err = fmt.Sscanf(line, "[%d-%d-%d %d:%d] wakes up",
		&e.Year, &e.Month, &e.Day, &e.Hour, &e.Minute)
	if err == nil {
		e.GuardId = 0
		e.Type = WAKES_UP
		return e, nil
	}

	return e, err
}

type Guard struct {
	Id              int
	Minutes         [60]int
	TotalAsleep     int
	MaxAsleep       int
	MaxAsleepMinute int
}

func (guard *Guard) Compute() {
	guard.TotalAsleep = 0
	guard.MaxAsleep = 0
	guard.MaxAsleepMinute = -1
	for i, m := range guard.Minutes {
		guard.TotalAsleep += m
		if guard.MaxAsleep < m {
			guard.MaxAsleep = m
			guard.MaxAsleepMinute = i
		}
	}
}

func (guard *Guard) Fill(start, end int) {
	for i := start; i < end; i++ {
		guard.Minutes[i]++
	}
}

func main() {
	input, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer input.Close()

	events := []Event{}
	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		line := scanner.Text()
		e, err := GetEvent(line)
		if err != nil {
			panic(err)
		}

		events = append(events, e)
	}

	sort.Slice(events, func(i, j int) bool {
		if events[i].Year != events[j].Year {
			return events[i].Year < events[j].Year
		}
		if events[i].Month != events[j].Month {
			return events[i].Month < events[j].Month
		}
		if events[i].Day != events[j].Day {
			return events[i].Day < events[j].Day
		}
		if events[i].Hour != events[j].Hour {
			return events[i].Hour < events[j].Hour
		}
		return events[i].Minute < events[j].Minute
	})

	guard_map := map[int]*Guard{}
	lastState := BEGINS_SHIFT
	guardId := 0
	minute := 0
	for _, e := range events {
		switch e.Type {
		case BEGINS_SHIFT:
			if guardId != e.GuardId && lastState == FALLS_ASLEEP {
				guard_map[guardId].Fill(minute, 60)
			}
			guardId = e.GuardId
			_, ok := guard_map[guardId]
			if !ok {
				guard_map[guardId] = &Guard{Id: guardId}
			}

		case FALLS_ASLEEP:
			minute = e.Minute

		case WAKES_UP:
			guard_map[guardId].Fill(minute, e.Minute)
		}
		lastState = e.Type
	}

	guards := []*Guard{}
	for _, guard := range guard_map {
		guard.Compute()
		guards = append(guards, guard)
	}

	sort.Slice(guards, func(i, j int) bool {
		return guards[i].TotalAsleep > guards[j].TotalAsleep
	})
	fmt.Println("Strategy1", guards[0].Id*guards[0].MaxAsleepMinute)

	sort.Slice(guards, func(i, j int) bool {
		return guards[i].MaxAsleep > guards[j].MaxAsleep
	})
	fmt.Println("Strategy2", guards[0].Id*guards[0].MaxAsleepMinute)
}
